File: Drivers/Leopard-Imaging-Camera/m021v4l2/opencv/cpp/flowcap.cpp
/*
   flowcap.cpp - Optical Flow from image capture  using Leopard Imageing M021 camear on Linux.
  
   Copyright (C) 2016 Simon D. Levy

   This file is part of M021_V4L2.

   M021_V4L2 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with M021_V4L2.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <opencv2/video/tracking.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

#include <iostream>

using namespace cv;
using namespace std;

#include "m021_v4l2_opencv.hpp"

#include <stdio.h>
#include <sys/timeb.h>

static const int   SCALEDOWN   = 1;

static const double PYRSCALE   = 0.5;
static const int    LEVELS     = 3;
static const int    WINSIZE    = 10;
static const int    ITERATIONS = 2;
static const int    POLY_N     = 2;
static const double POLY_SIGMA = 1.2;

// http://codepad.org/qPsNtwzp
static int getMilliCount(void){

    timeb tb;
    ftime(&tb);
    int nCount = tb.millitm + (tb.time & 0xfffff) * 1000;
    return nCount;
}

static void drawOptFlowMap(const Mat& flow, Mat& cflowmap, int step, const Scalar& color)
{
    double fx = 0;
    double fy = 0;

    for(int y = 0; y < cflowmap.rows; y += step)
        for(int x = 0; x < cflowmap.cols; x += step)
        {
            const Point2f& fxy = flow.at<Point2f>(y, x);
            fx += fxy.x;
            fy += fxy.y;
            line(cflowmap, Point(x,y), Point(cvRound(x+fxy.x), cvRound(y+fxy.y)), color);
            circle(cflowmap, Point(x,y), 2, color, -1);
        }

    printf("%4d %4d\r", (int)fx, (int)fy);
    fflush(stdout);
}

int main(int argc, char** argv)
{
    Mat flow, frame, frame2, gray, prevgray;
    
    //M021_800x460_Capture cap(frame);
    M021_800x460_Capture cap(frame);

    int start = getMilliCount();
    int flowcount = 0;

    while (true) {

        resize(frame, frame2, Size(frame.cols>>SCALEDOWN, frame.rows>>SCALEDOWN));

        cvtColor(frame2, gray, COLOR_BGR2GRAY);

        if(!prevgray.empty() ) {
        
            calcOpticalFlowFarneback(prevgray, gray, flow, 
                PYRSCALE, LEVELS, WINSIZE, ITERATIONS, POLY_N, POLY_SIGMA, 0);
            drawOptFlowMap(flow, frame2, 16, Scalar(0, 255, 0));
            imshow("flow", frame2);
            flowcount++;
        }

        if(waitKey(1)>=0)
            break;

        std::swap(prevgray, gray);
    }

    double duration = (getMilliCount() - start) / 1000.;

    int capcount = cap.getCount();

    int count = capcount > flowcount ? flowcount : capcount;

    printf("%d (%d) frames in %3.3f seconds = %3.3f frames /sec \n", count, flowcount, duration, count/duration);

    return 0;
}

File: Drivers/Leopard-Imaging-Camera/m021v4l2/opencv/python3/flowcap.py
#!/usr/bin/env python3

'''
   flowcap.py - Optical Flow from image capture  using Leopard Imageing M021 camear on Linux.
  
   Copyright (C) 2016 Simon D. Levy

   This file is part of M021_V4L2.

   M021_V4L2 is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
   BreezySTM32 is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with M021_V4L2.  If not, see <http://www.gnu.org/licenses/>.
'''

SCALEDOWN  = 2
PYRSCALE   = 0.5
LEVELS     = 3
WINSIZE    = 10
ITERATIONS = 2
POLY_N     = 2
POLY_SIGMA = 1.2
MOVESTEP   = 16
LINE_BGR   = 0,255,0

import cv2
from m021v4l2 import Capture800x460
from time import time

if __name__ == '__main__':

    cap = Capture800x460()

    start = time()
    flowcount = 0
    prev_gray = None
    flow = None

    newsize = 800>>SCALEDOWN, 460>>SCALEDOWN

    while True:

        # Capture frame-by-frame
        ret, frame = cap.read()

        frame2 = cv2.resize(frame, newsize)

        gray = cv2.cvtColor(frame2, cv2.COLOR_BGR2GRAY)

        if prev_gray is not None:

            flow = cv2.calcOpticalFlowFarneback(prev_gray, gray, flow,
                    pyr_scale=PYRSCALE, levels=LEVELS, winsize=WINSIZE, iterations=ITERATIONS, 
                    poly_n=POLY_N, poly_sigma=POLY_SIGMA, flags=0) 

            for y in range(0, flow.shape[0], MOVESTEP):
                for x in range(0, flow.shape[1], MOVESTEP):
                    fx, fy = flow[y, x]
                    cv2.line(frame2, (x,y), (int(x+fx),int(y+fy)), LINE_BGR)
                    cv2.circle(frame2, (x,y), 1, LINE_BGR, -1)

        cv2.imshow('LI-USB30-M021', frame2)
        if cv2.waitKey(1) & 0xFF == 27:
            break

        prev_gray = gray
        flowcount += 1

    duration = time() - start

    capcount = cap.getCount()

    count = flowcount if capcount > flowcount else capcount

    print('%d (%d) frames in %3.3f seconds = %3.3f frames /sec' % (count, flowcount, duration, count/duration))

